### PR TITLE
ref(ts): Project dashboard chart

### PR DESCRIPTION
**Pull request by scttcper**

### CODE REVIEW

File: src/sentry/static/sentry/app/views/settings/project/projectKeys/details/keyStats.tsx [hunk 1]
```diff
@@ -23,7 +23,16 @@ type Props = {
   'params'
 >;
 
-const getInitialState = () => {
+type State = {
+  since: number;
+  until: number;
+  loading: boolean;
+  error: boolean;
+  stats: any;
```
[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/static/sentry/app/views/settings/project/projectKeys/details/keyStats.tsx:31]: A TODO(ts) would be good to have.

File: src/sentry/static/sentry/app/components/barChart.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  markers: PropTypes.arrayOf(
+    PropTypes.shape({
+      x: PropTypes.number.isRequired,
+      label: PropTypes.string,
+    })
+  ),
+};
+
+BarChart.defaultProps = {
+  points: [],
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/barChart.tsx:37]: I used also to do that, but according to our docs we should use default params as the `defaultProps` on function components are going to be discontinued in the future 😉

https://develop.sentry.dev/frontend/defaultprops/#function-components

File: src/sentry/static/sentry/app/components/stackedBarChart.tsx [hunk 3]
```diff
@@ -11,7 +11,56 @@ import {use24Hours, getTimeFormat} from 'app/utils/dates';
...
+type Points = Point[];
+type Series = Array<{
+  data: Array<{x: number; y: number}>;
+  label?: string;
+  color?: string;
+}>;
+type Marker = {
+  x: number;
+  label?: string;
+  className?: string;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stackedBarChart.tsx:24]: I believe that we usually place className in the last position of the type object :)

File: src/sentry/static/sentry/app/components/stackedBarChart.tsx [hunk 4]
```diff
@@ -102,18 +149,18 @@ class StackedBarChart extends React.Component {
     }
   }
 
-  shouldComponentUpdate(nextProps, _nextState) {
+  shouldComponentUpdate(nextProps: Props, _nextState: State): boolean {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stackedBarChart.tsx:152]: I think you don't need to declare the return type in the life cycle methods

File: src/sentry/static/sentry/app/components/stackedBarChart.tsx (hunk 3 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/stackedBarChart.tsx]: moved!

File: src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx [hunk 5]
```diff
@@ -2,27 +2,32 @@ import React from 'react';
...
-      </div>
-    );
-  }
-}
+Chart.propTypes = {
+  stats: PropTypes.array,
+};
+
+Chart.defaultProps = {
+  stats: [],
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx:27]: Should we move it to the stats parameter?

File: src/sentry/static/sentry/app/views/settings/project/projectKeys/details/keyStats.tsx [hunk 6]
```diff
@@ -23,7 +24,17 @@ type Props = {
   'params'
 >;
 
-const getInitialState = () => {
+type State = {
+  since: number;
+  until: number;
+  loading: boolean;
+  error: boolean;
+  // TODO(ts): Add types to stats
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/project/projectKeys/details/keyStats.tsx:32]: 👏

File: src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx [hunk 7]
```diff
@@ -2,27 +2,32 @@ import React from 'react';
...
-  static defaultProps = {
-    stats: [],
-  };
+const Chart = ({stats}: Props) => {
+  const data = stats.map(d => ({x: d[0], y: d[1]}));
 
-  render() {
-    const data = this.props.stats.map(d => ({x: d[0], y: d[1]}));
+  return (
+    <div>
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx:16]: do you really need this div?

File: src/sentry/static/sentry/app/components/stackedBarChart.tsx [hunk 8]
```diff
@@ -50,26 +99,24 @@ class StackedBarChart extends React.Component {
...
-  constructor(props) {
+  constructor(props: Props) {
     super(props);
 
     // massage points
     let series = this.props.series;
 
-    if (this.props.points.length) {
-      if (series.length) {
+    if (this.props.points?.length) {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stackedBarChart.tsx:118]: here you could use the constructor `props`

**Review by priscilawebdev — COMMENTED** [review]
Almost there 😉

File: src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx (hunk 5 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx]: whoops thought i got both of them

File: src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx (hunk 7 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/views/projectsDashboard/chart.tsx]: removed div

File: src/sentry/static/sentry/app/components/stackedBarChart.tsx [hunk 9]
```diff
@@ -271,14 +318,20 @@ class StackedBarChart extends React.Component {
     );
   };
 
-  getMinHeight(index, _pointLength) {
+  getMinHeight(index: number, _pointLength: number): number {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stackedBarChart.tsx:321]:
```suggestion
  getMinHeight(index: number): number {
```

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/stackedBarChart.tsx:321]: as the second parameter is not being used, I would remove it from the function and here https://github.com/getsentry/sentry/blob/424938bc4d19b0897b87de063541a264e821a2f4/src/sentry/static/sentry/app/components/stackedBarChart.tsx#L344

**Review by priscilawebdev — APPROVED** [review]
Good Job! It looks good 👏

**Pull request merged**